wasm-node/src/node.rs

```diff
@@ -0,0 +1,179 @@
+            genesis_hash: config.genesis_hash,
+            p2p_local_keypair: config.p2p_local_keypair,
+            p2p_bootnodes: config.p2p_bootnodes,
+            p2p_listen_on: vec![],
+            store,
+        })
+        .await
+        .context("Failed to start node")
+        .unwrap_throw();
+
+        Self { node }
+    }
+
+    pub fn local_peer_id(&self) -> String {
+        self.node.p2p().local_peer_id().to_string()
+    }
+
+    pub async fn wait_connected(&self) {
+        self.node.p2p().wait_connected().await.unwrap_throw();
+    }
+
+    pub async fn wait_connected_trusted(&self) {
+        self.node
+            .p2p()
+            .wait_connected_trusted()
+            .await
+            .unwrap_throw();
+    }
+
+    pub async fn network_info(&self) -> JsValue {
+        // NetworkInfo isn't serializable
+        //let info = self.node.p2p().network_info().await.unwrap_throw();
+        //to_value(&info).unwrap_throw()
+        unimplemented!()
+    }
+
+    pub async fn get_head_header(&self) -> JsValue {
+        let eh = self.node.p2p().get_head_header().await.unwrap_throw();
+        to_value(&eh).unwrap_throw()
+    }
+
+    pub async fn get_header(&self, hash: &str /*&Hash*/) -> JsValue {
+        let hash = hash.parse().unwrap_throw();
+        let eh = self.node.p2p().get_header(hash).await.unwrap_throw();
+        to_value(&eh).unwrap_throw()
+    }
+
+    pub async fn get_header_by_height(&self, height: u64) -> JsValue {
+        let eh = self
+            .node
+            .p2p()
+            .get_header_by_height(height)
+            .await
+            .unwrap_throw();
+        to_value(&eh).unwrap_throw()
+    }
+
+    pub async fn get_verified_headers_range() {
+        unimplemented!()
+    }
+
+    pub async fn listeners(&self) -> Array {
+        self.node
+            .p2p()
+            .listeners()
+            .await
+            .unwrap_throw()
+            .iter()
+            .map(ToString::to_string)
+            .map(JsValue::from)
+            .collect::<Array>()
+    }
+
+    pub async fn connected_peers(&self) -> Array {
+        self.node
+            .p2p()
+            .connected_peers()
+            .await
+            .unwrap_throw()
+            .iter()
+            .map(ToString::to_string)
+            .map(JsValue::from)
+            .collect::<Array>()
+    }
+
+    pub async fn syncer_info(&self) -> JsValue {
+        let syncer_info = self.node.syncer().info().await.unwrap_throw();
+        to_value(&syncer_info).unwrap_throw()
+    }
+}
+
+#[wasm_bindgen]
+impl WasmNodeConfig {
+    #[wasm_bindgen(constructor)]
+    pub fn new(network: Network, genesis_hash: JsString, bootnodes: Vec<JsString>) -> Self {
+        let p2p_local_keypair = identity::Keypair::generate_ed25519();
+
+        let p2p_bootnodes: Vec<Multiaddr> = bootnodes
+            .iter()
+            .map(|addr| addr.as_string().unwrap_throw().parse().unwrap_throw())
```

Comment:
I think returning the `Result<T, JsValue>` works the same and we could use `?` too, but more type conversions could be needed

Reply:
@fl0rek could you check if we could have the `-> Result<T, JsValue>` instead of those unwrap throws? they scary me a bit :D

here we could do eg:
```suggestion
            .map(|addr| addr.as_string().and_then(|val| val.parse()))
            .collect::<Result<_, _>()?;
```

Reply:
this part of code got refactored, but to the original question - both `unwrap_throw` and returning `Result` give similar effect - from the js pointview, exception is thrown. Result way requires type conversion, but is slightly more friendly as we can control the error getting passed, for the former it's just "Uncaught Error: `unwrap_throw` failed" with a backtrace. I'll switch to `Result` where possible.

Reply:
Found additional resources about passing errors from Rust to Js:
https://github.com/rustwasm/wasm-bindgen/issues/1742
https://github.com/rustwasm/wasm-bindgen/issues/2970
So I've ended up defining `Result<T,`[`JsError`](https://docs.rs/wasm-bindgen/latest/wasm_bindgen/struct.JsError.html)`>`, which plays nicely with existing rust errors. Long term we could export our error type into js for more granular control.